### DISCUSSION TITLE
How to change keyboard language and change Japanese input method

### DISCUSSION THREAD

**Discussion by rlesandiego** [microsoft/microsoft-ui-xaml#7740] [Q&A]
I am working on an application that supports multiple languages and it allows the user to different languages.
Is there any way we can do this programmatically (via C#)?

Is there an equivalent to this in WPF like the following?

InputLanguageManager.Current.CurrentInputLanguage = new CultureInfo("ja-JP);

Is there any event to detect input language change?

Also, how do we change Japanese input mode and input method programmatically?

**Comment by castorix** [comment]
You can use [ActivateKeyboardLayout](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-activatekeyboardlayout) (tested on Windows 10 21H1)

**Comment by rlesandiego** [comment]
> 

ActivateKeyboardLayout() does not help. I want to be able to deterministically switch to a specific language, not to cycle the language list. Also, there is a need to programmatically  select a specific input method and input mode (Kana/Romaj input, Hiragana, Katakana, Alphanumeric, full-width, half-width) for Japanese (similarly to what  the language bar provides).

**Comment by castorix** [comment]
> ActivateKeyboardLayout() does not help. I want to be able to deterministically switch to a specific language, not to cycle the language list. 

You don't need to cycle
As the doc says, use _LoadKeyboardLayout_ first.

**Comment by rlesandiego** [comment]
> You can use [ActivateKeyboardLayout](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-activatekeyboardlayout) (tested on Windows 10 21H1)

This Win32 interface is not available in PInvoke.User32 package for Win UI3.

**Comment by castorix** [comment]
> This Win32 interface is not available in PInvoke.User32 package for Win UI3.

Why using a package ?
Just define it from the .h, like for any Win32 Api (or from .NET source code...)

```
        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern IntPtr ActivateKeyboardLayout(IntPtr hkl, uint Flags);

        public const int HKL_NEXT = 1;
        public const int HKL_PREV = 0;

        public const int KLF_ACTIVATE = 0x00000001;
        public const int KLF_SUBSTITUTE_OK = 0x00000002;
        public const int KLF_REORDER = 0x00000008;
        public const int KLF_REPLACELANG = 0x00000010;
        public const int KLF_NOTELLSHELL = 0x00000080;
        public const int KLF_SETFORPROCESS = 0x00000100;
        public const int KLF_SHIFTLOCK = 0x00010000;
        public const int KLF_RESET = 0x40000000;
```

**Comment by rlesandiego** [comment]
A WinUI3 project cannot directly refer to User32.dll. That's why PInvoke.User32 package was created.

**Comment by rlesandiego** [comment]
What does needs-triage label mean?
